jib-core/src/main/java/com/google/cloud/tools/jib/cache/Retry.java

```diff
@@ -0,0 +1,159 @@
+ */
+public class Retry<E extends Exception> {
+
+  /** An Action may throw an exception of type {@code E}. */
+  @FunctionalInterface
+  public interface Action<E extends Exception> {
+    void run() throws E;
+  }
+
+  /**
+   * Create a retryable action.
+   *
+   * @param action the action to be run
+   * @return the instance
+   * @param <E> the class of exceptions that may be thrown
+   */
+  public static <E extends Exception> Retry<E> action(Action<E> action) {
+    return new Retry<E>(action);
+  }
+
+  private final Action<E> action;
+  @Nullable private BooleanSupplier successCondition = null;
+  private int maximumRetries = 5;
+  private Predicate<? super E> retryOnException = ignored -> true; // continue to retry
+  private long sleepMilliseconds = -1; // no sleep
+
+  private Retry(Action<E> action) {
+    this.action = action;
+  }
+
+  /**
+   * Set the success condition: the action will be retried until this condition is true.
+   *
+   * @param successCondition the stop condition
+   * @return the instance for further configuration
+   */
+  public Retry<E> until(BooleanSupplier successCondition) {
+    this.successCondition = successCondition;
+    return this;
+  }
+
+  public Retry<E> maximumRetries(int maximumRetries) {
+    Preconditions.checkArgument(maximumRetries > 0);
+    this.maximumRetries = maximumRetries;
+    return this;
+  }
+
+  /**
+   * Provide a predicate to determine if a thrown exception can be retried.
+   *
+   * @param retryOnException determine if provided exception is retryable.
+   * @return the instance for further configuration
+   */
+  public Retry<E> retryOnException(Predicate<Exception> retryOnException) {
+    this.retryOnException = retryOnException;
+    return this;
+  }
+
+  /**
+   * Set the sleep time between retries.
+   *
+   * @param duration the time to sleep
+   * @param unit the unit of time of duration
+   * @return the instance for further configuration
+   */
+  public Retry<E> sleep(long duration, TimeUnit unit) {
+    Preconditions.checkArgument(duration > 0);
+    this.sleepMilliseconds = unit.convert(duration, TimeUnit.MILLISECONDS);
+    return this;
+  }
+
+  /**
+   * Run the action until it runs successfully, to a {@link #maximumRetries(int) maximum number of
+   * retries} (default: 5). An action is deemed to be successful if there are no exceptions and the
+   * {@link #until(BooleanSupplier) optional success condition} is true. If an exception occurs then
+   * the action will be retried providing {@link #retryOnException(Predicate) the exception is
+   * retryable}.
+   *
+   * @return true if the action was run successfully, or {@code false} if the action was unable to
+   *     complete
+   * @throws E exception thrown during the action
+   */
+  public boolean run() throws E {
+    for (int i = 0; i < maximumRetries; i++) {
+      // sleep between attempts, but not on the first attempt
+      if (i > 0 && sleepMilliseconds >= 0) {
```

Comment:
`sleepMilliseconds > 0` instead of `>=`?

Reply:
Oops, I meant for 0 as it (almost) acts like a yield. Will fix `sleep()` instead.